src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +404,202 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+      break;
+    }
+
+    struct ArrowBufferView src_view = src->buffer_views[i];
+    int64_t buffer_size = src_view.size_bytes;
+
+    // No need to initialize a buffer of zero bytes
+    if (buffer_size == 0) {
+      continue;
+    }
+
+    // buffer_size will be -1 if it is unknown (i.e., string/binary
+    // data buffer when copying from CUDA -> CPU)
+    if (buffer_size < 0) {
+      (*unknown_buffer_size_count)++;
+      continue;
+    }
+
+    struct ArrowBuffer* buffer_dst = ArrowArrayBuffer(dst, i);
+
+    // Don't initialize a buffer if we have already started the copy.
+    // This check works because the buffer size is set synchronously
+    // when the copy operation is launched.
+    if (buffer_dst->size_bytes > 0) {
+      continue;
+    }
+
+    // Otherwise, initialize the buffer
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaBufferInitAsync(
+        device_src, src_view, device_dst, buffer_dst, stream));
+  }
+
+  // Recurse for children
+  for (int64_t i = 0; i < src->n_children; i++) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->children[i], device_dst, dst->children[i], stream,
+        unknown_buffer_size_count));
+  }
+
+  // ...and for dictionary
+  if (src->dictionary != NULL) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->dictionary, device_dst, dst->dictionary, stream,
+        unknown_buffer_size_count));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaCreateStream(struct ArrowDevice* src,
+                                                  struct ArrowDevice* dst, CUstream* out,
+                                                  struct ArrowError* error) {
+  struct ArrowDeviceCudaPrivate* private_data;
+
+  if (dst->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)dst->private_data;
+  } else if (src->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)src->private_data;
+  } else {
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuCtxPushCurrent(private_data->cu_context),
+                               "cuCtxPushCurrent", NULL);
+  CUresult err = cuStreamCreate(out, CU_STREAM_NON_BLOCKING);
+  CUcontext unused;
+  NANOARROW_CUDA_ASSERT_OK(cuCtxPopCurrent(&unused));
+  NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamCreate", error);
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaArrayViewCopyInternal(
+    struct ArrowDevice* device_src, struct ArrowDeviceArrayView* src,
+    struct ArrowDevice* device_dst, struct ArrowArray* dst, CUstream* stream,
+    struct ArrowError* error) {
+  CUresult err;
+  int code;
+  int64_t unknown_buffer_size_count = 0;
+
+  NANOARROW_RETURN_NOT_OK_WITH_ERROR(
+      ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst, dst,
+                                          stream, &unknown_buffer_size_count),
+      error);
```

Comment:
You probably need to call `cuStreamWaitEvent` if there's a source event before you pass it to these functions